keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt

```diff
@@ -157,12 +157,15 @@ class ResourceActuator(
           }
           else -> {
             log.info("Resource {} is valid [$coid]", id)
-            // TODO: not sure this logic belongs here
             val lastEvent = resourceRepository.lastEvent(id)
-            if (lastEvent is ResourceDeltaDetected || lastEvent is ResourceActuationLaunched) {
-              publisher.publishEvent(ResourceDeltaResolved(resource, clock))
-            } else {
-              publisher.publishEvent(ResourceValid(resource, clock))
+            when (lastEvent) {
+              is ResourceActuationLaunched -> log.debug("waiting for actuating task to be completed") // do nothing and wait
+              is ResourceDeltaDetected, is ResourceTaskSucceeded, is ResourceTaskFailed -> {
+                // if a delta was detected and a task wasn't launched, the delta is resolved
+                // if a task was launched and it completed, either successfully or not, the delta is resolved
+                publisher.publishEvent(ResourceDeltaResolved(resource, clock))
```

Comment:
I asked this question in another comment, but I'm confused as to how a delta would have been resolved if the task launched to resolve the delta has failed...

Reply:
So this case is only hit when there is no delta (aka the resource is in a good state). It would be weird if this happened (maybe the task timed out? maybe it failed on some other step like scaling down an old server group?) but regardless there was no diff detected, so we should output the right event.

All the other events will still stay in the task history.

Reply:
Sorry, I don't follow. If a task was launched and failed it's because there was a delta to resolve, no? Why do you say this is only hit when there is no delta?

Reply:
Funny enough, Emily's example case has happened to me multiple times! Both times the scenario has been:
1. Delta detected
2. Deploy task is launched
3. Server group comes up healthy and fine, but something about the deploy made the task itself fail (one time it was emily's timeout example, the other time it was something wonky with the disable step)
4. There's now a failed task in the history events
5. Keel comes back around and sees things are back to the right state, and considers things resolved, it adds a `ResourceValid` event

The relationship between orca's view of the world (did this unit of work go the way i expected) and keel's view of the world (is the current state of this thing the same as how i'd ideally like it to be) are very different in areas like this.

Reply:
@luispollo the `ResourceTaskFailed` event is connected to orca tasks. When there's a diff, and there's an actuation, the task can either succeed or failed. In both cases. the diff is resolved.

Reply:
Thanks, @erikmunson. That clarified it for me! @gal-yardeni, that's not entirely true. If an actuation task is launched (say, for a deployment), and the deployment itself fails, the task will be marked as failed, and the diff will _not_ have been resolved because current state will differ from desired state. However, in this case, the block of code you're in is only reached when we've already compared and there is no diff, so in this specific spot it's safe to make that assumption.

I think what threw me off was reading the comment just above these lines, without realizing that we were in this `else` block that's only reached when `current == desired`.